pkg/gatewaysecret/handler.go

```diff
@@ -0,0 +1,142 @@
+package gatewaysecret
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	certmanagerv1 "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	"github.com/go-logr/logr"
+	apicorev1 "k8s.io/api/core/v1"
+	apimetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/kyma-project/lifecycle-manager/pkg/util"
+)
+
+const (
+	LastModifiedAtAnnotation = "lastModifiedAt"
+	gatewaySecretName        = "klm-istio-gateway" //nolint:gosec // gatewaySecretName is not a credential
+	kcpRootSecretName        = "klm-watcher"
+	kcpCACertName            = "klm-watcher-serving"
+	istioNamespace           = "istio-system"
+)
+
+var errCouldNotGetLastModifiedAt = errors.New("getting lastModifiedAt time failed")
+
+type GatewaySecretHandler struct {
+	SecretManager
+}
+
+func NewGatewaySecretHandler(secretManager SecretManager) *GatewaySecretHandler {
+	return &GatewaySecretHandler{
+		secretManager,
+	}
+}
+
+func (gsh *GatewaySecretHandler) ManageGatewaySecret(ctx context.Context, rootSecret *apicorev1.Secret) error {
+	gwSecret, err := gsh.FindGatewaySecret(ctx)
+
+	if util.IsNotFound(err) {
+		return gsh.handleNonExisting(ctx, rootSecret)
+	}
+	if err != nil {
+		return err
+	}
+
+	return gsh.handleExisting(ctx, rootSecret, gwSecret)
+}
+
+func (gsh *GatewaySecretHandler) handleNonExisting(ctx context.Context, rootSecret *apicorev1.Secret) error {
+	gwSecret := NewGatewaySecret(rootSecret)
+	return gsh.Create(ctx, gwSecret)
+}
+
+func (gsh *GatewaySecretHandler) handleExisting(ctx context.Context,
+	rootSecret *apicorev1.Secret, gwSecret *apicorev1.Secret,
+) error {
+	caCert, err := gsh.GetRootCACertificate(ctx)
+	if err != nil {
+		return err
+	}
+
+	if copied := CopyRootSecretDataIntoGatewaySecret(gwSecret, caCert, rootSecret); !copied {
+		return nil
+	}
+
+	return gsh.Update(ctx, gwSecret)
+}
+
+func CopyRootSecretDataIntoGatewaySecret(gwSecret *apicorev1.Secret, caCert certmanagerv1.Certificate,
+	rootSecret *apicorev1.Secret,
+) bool {
+	if !GatewaySecretRequiresUpdate(gwSecret, caCert) {
```

Comment:
Following the "Single Responsibility Principle" I don't think putting this condition here is a good choice.
First, it complicates the unit test: You have to setup some timestamps just to test if "copy" works...
Second, it complicates the signature of this function, `caCert` is used only to test if copying is required.
I would move this check to the place where `CopyRootSecretDataIntoGatewaySecret` is invoked (i.e. the body of the `handleExisting` method)